### PR TITLE
[engsys] ci: Fix `tox_harness.py` dropping package extras when rewriting dev_requirements.txt

### PR DESCRIPTION
**Pull request by kdestin**
# Description

This pull request fixes an issue where `scripts/devops_tasks/tox_harness.py` will rewrite requirements without any specified extras (`./path/to/package[extra,extra1,...]` -> `/path/to/package.whl`)

This pull request helps unblock #32770 

**Note**: This pull request may have the potential to cause CI failures. CI that were in a previously passing state with package extras dropped may have different behavior/fai

* Example: A couple requirements files install `azure-sdk-tools` with the `build` extra

https://github.com/Azure/azure-sdk-for-python/blob/e3bb7f2aafb10b1f87af735ce7fdf3b031b42786/eng/ci_tools.txt#L39

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: scripts/devops_tasks/tox_harness.py [hunk 1]
```diff
@@ -119,17 +119,18 @@ def replace_dev_reqs(file, pkg_root):
...
+                if extras:
+                    extras = f"[{extras}"
 
-            adjusted_req_lines.append(amended_line)
+            adjusted_req_lines.append(f"{build_whl_for_req(amended_line, pkg_root)}{extras}")
 
     req_file_name = os.path.basename(file)
     logging.info("Old {0}:{1}".format(req_file_name, adjusted_req_lines))
 
-    adjusted_req_lines = list(map(lambda x: build_whl_for_req(x, pkg_root), adjusted_req_lines))
```
Comment (scbedd) [inline scripts/devops_tasks/tox_harness.py:132]: While I'd factor it differently now, what I was thinking when originally writing this code was to give a place where we print the lines before and after we adjust them. In this case, if you look at the log of `old` and `new` they will be the same if you pull this into the loop.  I'd like to keep that around, I'm betting this is how you located that those dev reqs were being replaced.

**Review by scbedd — CHANGES_REQUESTED** [review]
Agree with the bugfix, just would prefer it factored a tiny bit differently.

File: scripts/devops_tasks/tox_harness.py (hunk 1 above)
Comment (kdestin) [reply scripts/devops_tasks/tox_harness.py:132]: My bad, I thought I'd kept that behavior as it was!

---

Tentatively resolved by: https://github.com/Azure/azure-sdk-for-python/pull/32864/commits/eae1d0cb58dc6097d297c8b61227f32ba883cf21

The behavior is _slightly_ different. The "old" log message wouldn't show any `-e` that were in the original requirement. But the above commit preserves them.

Comment (scbedd) [reply scripts/devops_tasks/tox_harness.py:132]: Honestly that's better, so I won't complain. 👍

**Review by scbedd — APPROVED** [review]
Looks great, thanks muchly @kdestin

**Comment by scbedd** [comment]
I am going to check-enforcer override this. Looks like we don't have `template` set up to trigger on `tox_harness` changes. Will fix that with my next pass through this code.

For folks observing this PR,  [manually queued](https://dev.azure.com/azure-sdk/public/_build/results?buildId=3215951&view=logs&j=84b20eb0-323b-5722-f041-399ea9857d18&t=1024ff5b-59a4-5548-1c1b-bbb298ca1664) the core public CI build against this change. Everything looks to be checking out properly.

**Comment by scbedd** [comment]
The test failure is an existing issue that I'm chasing. Ignoring for merge.

**Comment by scbedd** [comment]
@microsoft-github-policy-service rerun

**Pull request merged**